src/sql/pool.rs

```diff
@@ -0,0 +1,105 @@
+//! Connection pool.
+
+use std::fmt;
+use std::ops::{Deref, DerefMut};
+use std::sync::{Arc, Weak};
+
+use parking_lot::{Condvar, Mutex};
+use rusqlite::Connection;
+
+/// Inner connection pool.
+struct InnerPool {
+    /// Available connections.
+    connections: Mutex<Vec<Connection>>,
+
+    /// Conditional variable to notify about added connections.
+    ///
+    /// Used to wait for available connection when the pool is empty.
+    cond: Condvar,
+}
+
+impl InnerPool {
+    /// Puts a connection into the pool.
+    ///
+    /// The connection could be new or returned back.
+    fn put(&self, connection: Connection) {
+        let mut connections = self.connections.lock();
+        connections.push(connection);
+        drop(connections);
+        self.cond.notify_one();
+    }
+}
+
+/// Pooled connection.
+pub struct PooledConnection {
+    /// Weak reference to the pool used to return the connection back.
+    pool: Weak<InnerPool>,
+
+    /// Only `None` right after moving the connection back to the pool.
+    conn: Option<Connection>,
```

Comment:
Ugly... Looks like not implementing the `Default` trait is generally an anti-pattern

Reply:
Also don't know why `drop()` takes a reference instead of a value. The problem with `drop()` recursion seems to be easy to solve at compiler level